### ISSUE TITLE
parallel performance of FDS6

### ISSUE THREAD

**Issue by gforney** [firemodels/fds#2084]
```
Please complete the following lines...

FDS Version: 6.0.1
SVN Revision Number: 17534
Compile Date: 26 Nov 2013
Smokeview Version/Revision:
Operating System: Windows Server 2012

Describe details of the issue below:

When splitting up a domain into multiple meshes to reduce calculation time, FDS6 does
not scale as well on multiple cores compared to FDS5.

I used a test case consisting of a 1 million cell empty box which was split up into
2, 4, 8, and 16 meshes for the various tests and ran for 10 seconds.  The graph attached
shows the speed up obtained for both FDS5 and FDS6.  I understand that FDS6 is slower
but in FDS5 when running upto 8 cores good speedup was obtained with the results flat
lining beyond this point.  With FDS6 beyond 4 cores the effect is reducing and by 16
cores the case is running for longer.

All cases were carried out on a dual cpu Xeon E5-2690 compute node with no other programmes
running.

FDS5 line:
mpiexec -n x -channel ssm -priority 1:3 

FDS6 line:
mpiexec -n x -channel nemesis -priority 1:3 

If you have a benchmark case that you normally use for testing parallel performance
I can re run using this case for consistency.

```

Original issue reported on code.google.com by `ajbasford` on 2014-04-01 14:46:02

<hr>
- _Attachment: number of cores comparrison.png<br>![number of cores comparrison.png](https://storage.googleapis.com/google-code-attachments/fds-smv/issue-2120/comment-0/number of cores comparrison.png)_

**Comment by gforney** [comment]
```
We have noticed that the OpenMP version of FDS 6 also has this trend. I cannot at the
moment explain why FDS 5 would be more efficient on multiple cores, but I think in
general that there is a point of diminishing returns running FDS in parallel (using
either MPI or OpenMP) on a single computer because there is only one block of RAM and
limited bandwidth getting data off the stack and onto the chip.

We may combine this issue with others related to parallel processing.
```

Original issue reported on code.google.com by `mcgratta` on 2014-04-01 15:47:03

**Comment by gforney** [comment]
```
Assuming that the submitted benchmark was performed using only MPI and not MPI-OpenMP
hybrid, these are two separate issues.

I cannot comment as to why the MPI version doesn't scale past 8 processes and why FDS6
even slows down past that point.

For OpenMP the issue is definitely memory bandwidth. You'll see significant differences
in scaling depending on your memory bandwidth. I'll be submitting some data for this
to the guide shortly.

This also means that Hyperthreading/SMT hurts performance. So once you push your OpenMP
threadcount past the physical core count you won't see any more improvements or even
a slowdown. I'm no expert but my understanding is that SMT will help you when one thread
can perform calculations while the other one is waiting for data from memory. Since
most of FDS appears to be severely memory bound there aren't enough cycles to occupy
the second thread. So we end up with both waiting for data.
```

Original issue reported on code.google.com by `daniel.haarhoff` on 2014-04-02 06:30:56

**Comment by gforney** [comment]
```
Thanks for the responses.

I ran the test using only MPI.  Hyper threading is disabled on all cores so these are
all physical cores within one box.  Do you think its a RAM bottleneck issue?  Each
processor has 4 sticks of 8GB 1600 DDR3 RAM.  I can try comparing the effect of running
8 MPI threads on 1 processor only and 4 MPI threads on each processor as this would
double the memory bandwidth if you think this is the issue.
```

Original issue reported on code.google.com by `ajbasford` on 2014-04-02 10:54:19

**Comment by gforney** [comment]
```
No, I think the next step is running some more in depth profiling with Scalasca or similar
tools. Which I might get round to in the next weeks.
Your RAM only is an issue if it can't max out your processors memory bandwidth or your
simulation is too large to fit in RAM.

My guess right now would be that the issue has to do with the fact that your 16 cores
are split between two sockets. The bandwidth and latency between sockets is worse than
between cores within one socket.
So maybe FDS6 has increased communication demands compared to FDS5.

One thing you could do would be to distribute the processes differently. For example
four on each socket instead of eight on one socket.
```

Original issue reported on code.google.com by `daniel.haarhoff` on 2014-04-02 11:28:34
- **Labels added**: Type-Other
- **Labels removed**: Type-Defect

**Comment by gforney** [comment]
```
I have had a further look at this and I think the problem is the change in MPICH version
and the way it sets processor affinity.  I therefore added the following line to the
mpiexec command so it now reads:

mpiexec -n 16 -binding user:0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15 -channel nemesis
fds_mpi.exe job_name.fds

What this does is instead of letting windows randomly distributing each process across
the cpus, instead each process is locked to an individual core.  

This can clearly be seen when running on say 8 cores as now 8 processor graphs are
at maximum whereas previously the levels on all 16 would vary continuously and has
a significant effect on the speed of running the job.


```

Original issue reported on code.google.com by `ajbasford` on 2014-04-02 15:34:04

**Comment by gforney** [comment]
```
Great! Have you maybe also tried something like this:

mpiexec -n 16 -binding user:0,8,1,9,2,10,3,11 -channel nemesis fds_mpi.exe job_name.fds

If you could post your results as a csv that would be great.
```

Original issue reported on code.google.com by `daniel.haarhoff` on 2014-04-03 06:12:21

**Comment by gforney** [comment]
```
I have now tried what you said and updated the graph and attached a csv file.

Using the binding option has given scaling consistent with what was previously achieved
with FDS5.  When distributing the binding this gives a significant speedup on a socket
system unless n equals the number of physical cores in which case you are best sticking
with increasing the number linearly.

The graphs show the following:
FDS 5 runtime - Case ran in FDS5
FDS 5 runtime - Case ran in FDS6 with no binding options
Linear bindings used - Case ran in FDS6 with the bindings increasing linearly.  The
exact bindings used for each case are in the csv file.
Distributed bindings used - Case fan in FDS6 with the bindings distributed as your
post suggests.  The exact bindings used for each case are in the csv file.
Mixed bindings - This takes the optimal bindings settings from the other tests.



```

Original issue reported on code.google.com by `ajbasford` on 2014-04-03 10:25:27

<hr>
- _Attachment: number of cores comparrison 2.png<br>![number of cores comparrison 2.png](https://storage.googleapis.com/google-code-attachments/fds-smv/issue-2120/comment-7/number of cores comparrison 2.png)_
- _Attachment: [fds_runtimes.csv](https://storage.googleapis.com/google-code-attachments/fds-smv/issue-2120/comment-7/fds_runtimes.csv)_

**Comment by gforney** [comment]
```
This issue is now fixed in FDS v6.1.2 using the new intel mpi implementation as the
mpi processes are bound to individual cores by default.
```

Original issue reported on code.google.com by `ajbasford` on 2014-10-17 12:15:28

**Comment by gforney** [comment]
```
Great. Thanks for letting us know. I'll mark the Issue as Verified.
```

Original issue reported on code.google.com by `mcgratta` on 2014-10-17 12:54:29
